crates/testing-framework/src/ephemeral/mod.rs

```diff
@@ -0,0 +1,248 @@
+            default_neiborhood: default_neiborhoud.map(Into::into).collect(),
+        }
+    }
+
+    pub fn from_vec(test_parameters: TestRunParameters, nodes: Vec<Peer>) -> Self {
+        let mut network = Self::empty(test_parameters);
+        let neighborhood: PeerSet = nodes.iter().map(|peer| peer.peer_id.clone()).collect();
+        for peer in nodes {
+            // TODO can peer have itself as a neighbor?
+            network.add_peer_with_neighborhood(peer, neighborhood.iter().cloned());
+        }
+        network
+    }
+
+    pub fn add_peer_with_neighborhood(
+        &mut self,
+        peer: Peer,
+        neighborhood: impl IntoIterator<Item = impl Into<PeerId>>,
+    ) -> &mut PeerWithNeighborhood {
+        let peer_id = peer.peer_id.clone();
+        let mut peer_with_neigh = PeerWithNeighborhood::new(peer);
+        peer_with_neigh.extend_neighborhood(neighborhood.into_iter().map(Into::into));
+        self.peers
+            .insert(peer_id.clone(), Rc::new(peer_with_neigh.into()));
+        Rc::get_mut(self.peers.get_mut(&peer_id).unwrap())
+            .unwrap()
+            .get_mut()
+    }
+
+    /// Add a peer with default neighborhood.
+    pub fn add_peer(&mut self, peer: Peer) -> &mut PeerWithNeighborhood {
+        let peer_id = peer.peer_id.clone();
+        let mut peer_with_neigh = PeerWithNeighborhood::new(peer);
+        peer_with_neigh.extend_neighborhood(self.default_neiborhood.iter().cloned());
+        self.peers
+            .insert(peer_id.clone(), Rc::new(peer_with_neigh.into()));
+        Rc::get_mut(self.peers.get_mut(&peer_id).unwrap())
+            .unwrap()
+            .get_mut()
+    }
+
+    pub fn set_peer_failed<Id>(&mut self, peer_id: &Id, failed: bool)
+    where
+        PeerId: Borrow<Id>,
+        Id: Hash + Eq + ?Sized,
+    {
+        self.peers
+            .get_mut(peer_id)
+            .expect("unknown peer")
+            .as_ref()
+            .borrow_mut()
+            .set_failed(failed);
+    }
+
+    pub fn fail_peer_for<Id>(&mut self, source_peer_id: &Id, target_peer_id: impl Into<PeerId>)
+    where
+        PeerId: Borrow<Id>,
+        Id: Hash + Eq + ?Sized,
+    {
+        self.peers
+            .get_mut(source_peer_id)
+            .expect("unknown peer")
+            .as_ref()
+            .borrow_mut()
+            .get_neighborhood_mut()
+            .fail(target_peer_id);
+    }
+
+    pub fn unfail_peer_for<Id1, Id2>(&mut self, source_peer_id: &Id1, target_peer_id: &Id2)
+    where
+        PeerId: Borrow<Id1>,
+        Id1: Hash + Eq + ?Sized,
+        PeerId: Borrow<Id2>,
+        Id2: Hash + Eq + ?Sized,
+    {
+        self.peers
+            .get_mut(source_peer_id)
+            .expect("unknown peer")
+            .as_ref()
+            .borrow_mut()
+            .get_neighborhood_mut()
+            .unfail(target_peer_id);
+    }
+
+    // TODO there is some kind of unsymmetry between these methods and the fail/unfail:
```

Comment:
I think it depends on usecases. If it is useful to call fail/unfail on unknown peers then it is ok to just check if the peer is in the neighborhood. I personally like approach "it panics if you do ANYTHING wrong, assert any improper movement". But that can be overrides bu the goal of simplicity of use. We are here for the improvement of the  DX of the AquaVM, after all.